pkg/apis/serving/v1beta1/configmap.go

```diff
@@ -0,0 +1,159 @@
+type ExplainerConfig struct {
+	// explainer docker image name
+	ContainerImage string `json:"image"`
+	// default explainer docker image version
+	DefaultImageVersion string `json:"defaultImageVersion"`
+}
+
+// +kubebuilder:object:generate=false
+type ExplainersConfig struct {
+	AlibiExplainer ExplainerConfig `json:"alibi,omitempty"`
+}
+
+// +kubebuilder:object:generate=false
+type PredictorConfig struct {
+	// predictor docker image name
+	ContainerImage string `json:"image"`
+	// default predictor docker image version on cpu
+	DefaultImageVersion string `json:"defaultImageVersion"`
+	// default predictor docker image version on gpu
+	DefaultGpuImageVersion string `json:"defaultGpuImageVersion"`
+}
+
+// +kubebuilder:object:generate=false
+type PredictorsConfig struct {
+	Tensorflow PredictorConfig `json:"tensorflow,omitempty"`
+	Triton     PredictorConfig `json:"triton,omitempty"`
+	XGBoost    PredictorConfig `json:"xgboost,omitempty"`
+	SKlearn    PredictorConfig `json:"sklearn,omitempty"`
+	PyTorch    PredictorConfig `json:"pytorch,omitempty"`
+	ONNX       PredictorConfig `json:"onnx,omitempty"`
+}
+
+// +kubebuilder:object:generate=false
+type TransformerConfig struct {
+	// transformer docker image name
+	ContainerImage string `json:"image"`
+	// default transformer docker image version
+	DefaultImageVersion string `json:"defaultImageVersion"`
+}
+
+// +kubebuilder:object:generate=false
+type TransformersConfig struct {
+	Feast TransformerConfig `json:"feast,omitempty"`
+}
+
+// +kubebuilder:object:generate=false
+type InferenceServicesConfig struct {
+	// Transformer configurations
+	Transformers *TransformersConfig `json:"transformers"`
+	// Predictor configurations
+	Predictors *PredictorsConfig `json:"predictors"`
+	// Explainer configurations
+	Explainers *ExplainersConfig `json:"explainers"`
+}
+
+func GetInferenceServicesConfig(client client.Client) (*InferenceServicesConfig, error) {
+	configMap := &v1.ConfigMap{}
+	err := client.Get(context.TODO(), types.NamespacedName{Name: constants.InferenceServiceConfigMapName, Namespace: constants.KFServingNamespace}, configMap)
+	if err != nil {
+		return nil, err
+	}
+
+	inferenceServiceConfigMap, err := NewInferenceServicesConfig(configMap)
+	if err != nil {
+		return nil, err
+	}
+
+	return inferenceServiceConfigMap, nil
+}
+
+func NewInferenceServicesConfig(configMap *v1.ConfigMap) (*InferenceServicesConfig, error) {
+	predictorsConfig, err := getPredictorsConfigs(configMap)
+	if err != nil {
+		return nil, err
+	}
+	transformersConfig, err := getTransformersConfigs(configMap)
+	if err != nil {
+		return nil, err
+	}
+	explainersConfig, err := getExplainersConfigs(configMap)
+	if err != nil {
+		return nil, err
+	}
+	return &InferenceServicesConfig{
+		Predictors:   predictorsConfig,
+		Transformers: transformersConfig,
+		Explainers:   explainersConfig,
+	}, nil
+}
+
+func getPredictorsConfigs(configMap *v1.ConfigMap) (*PredictorsConfig, error) {
```

Comment:
You could abstract this for all 3 like
```
func getComponentConfig(config interface, key string) error {}
```
In fact, is there any reason you can't let the serialization library do all of this generically and just do
```
json.Unmarshal(InferenceServicesConfig, configmap.Data)
```

Reply:
We are not storing the entire config as one json, it is multiple data items in the configmap.

Reply:
Ah I see. configMap.Data is a map[string]string. Still, can be done generically like:
```
cfg: = InferenceServicesConfig{}
for _, err := []error{
  getComponentConfig(PredictorConfigKeyName, cfg.Predictors),
  getComponentConfig(ExplainerConfigKeyName, cfg.Explainers),
  getComponentConfig(TransformerConfigKeyName, cfg.Transformers),
} {
  if err != nil {
    return nil, err 
  } 
}
return cfg, nil

...
func getComponentConfig(key string, cfg interface) error { 
...
}
```
Just have to make sure InferenceServicesConfig doesn't use pointers, which it probably shouldn't anyways.

Reply:
The config structs for predictor, explainer and transformer are fairly different, not sure how we can use the cfg interface.

Reply:
The helper function doesn't do anything beyond pass the struct to ```err := json.Unmarshal([]byte(data), &predictorConfig)```. This function takes an interface https://golang.org/pkg/encoding/json/#Unmarshal, so you don't actually need to know what kind of config it is when you serialize it, it will use reflection to load the variables. 

If this doesn't compile, I may be confused about something in golang, but I'm 99% sure you can do this.

Reply:
I think you are right, fixed.